src/core/stores/VRange.ts

```diff
@@ -0,0 +1,94 @@
+import { VNode, VNodeType } from './VNode';
+
+export enum RangeDirection {
+    BACKWARD = 'BACKWARD',
+    FORWARD = 'FORWARD',
+}
+export enum RelativePosition {
+    BEFORE = 'BEFORE',
+    AFTER = 'AFTER',
+}
+export interface TargetLocation {
+    reference: VNode;
+    relativePosition: RelativePosition;
+}
+export interface VRangeLocation {
+    start: TargetLocation;
+    end: TargetLocation;
+    direction: RangeDirection;
+}
+
+export class VRange {
+    readonly _start = new VNode(VNodeType.RANGE_START);
+    readonly _end = new VNode(VNodeType.RANGE_END);
+    direction: RangeDirection;
+    constructor(direction: RangeDirection = RangeDirection.FORWARD) {
+        this.direction = direction;
+    }
+    /**
+     * Collapse the range. Return self.
+     *
+     * @param [onEnd] true to collapse on the end range
+     */
+    collapse(onEnd = false): VRange {
```

Comment:
Because it has virtually nothing to do with the range direction. The range direction will always be forward if the range is collapsed. This is about whether we want to collapse the range on its current start point or on its current _end_ point. But I do agree that a `boolean` is too cryptic. Also to note, in the vast majority of cases we will want to collapse on the start point.

Reply:
How about this ?
```typescript
/**
 * Collapse the range. Return self.
 *
 * @param [onEnd] true to collapse on the end range
 */
collapse(edge: 'start' | 'end' = 'start'): VRange {
    if (edge === 'start') {
        return this.setEnd(this._start);
    } else {
        return this.setStart(this._end);
    }
}
```